src/check_bgp_protocols.py

```diff
@@ -0,0 +1,138 @@
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import os
+import re
+import sys
+
+from argparse import ArgumentParser
+from subprocess import Popen, PIPE, STDOUT
+
+
+def parse_args():
+    parser = ArgumentParser(
+        description='Check if all BIRD BGP protocols are up and established'
+    )
+    parser.add_argument('path', help='Path to the birdc/birdc6 binary')
+
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    birdc_path = args.path
+
+    # Check if the path to birdc actually exists and is executable
+    if not check_birdc_executable(birdc_path):
+        print(format_nagios_message(
+            ExitCodes.critical,
+            'The path to the birdc executable does not exist/is not executable'
+        ))
+        sys.exit(ExitCodes.critical)
+
+    # Run the actual check and get the results
+    code, reason = check_birdc_peers(birdc_path)
+
+    print(format_nagios_message(code, reason))
+    sys.exit(code)
+
+
+def check_birdc_peers(path):
+    protocols = get_birdc_protocols(path)
+
+    if not protocols:
+        return (ExitCodes.critical, 'Could not retrieve BIRD\'s output')
+
+    # Here we match the BGP protocol lines from BIRD.
+    # Below is an example of how they look like:
+    # direct1  Direct   master   up     2020-02-27
+    # static_bgp Static   master   up     2020-02-21
+    # b_firstroot BGP      master   up     2020-02-28  Established
+    # b_lb     BGP      master   start  18:43:46    Active        Socket: conn
+    # b_af     BGP      master   up     2020-02-29  Established
+    # b_home   BGP      master   down   2020-02-29
+    # We are extracting the protocol name, state, date/time since last
+    # change and the additional information about the state of the session.
+    r = re.compile(
+        r'^(?P<name>\w+)\s+BGP\s+(\w+)\s+(?P<state>\w+)\s+'
+        r'(?P<since>[\d\-:]+)\s+(?P<info>[\w]+)?\s*$',
+        re.MULTILINE
+    )
+
+    res = [m.groupdict() for m in r.finditer(protocols)]
+
+    if not res:
+        return (ExitCodes.critical, 'Could not parse BIRD\'s output')
+
+    protocols_down = [
+        protocol['name'] for protocol in res
+        if protocol['info'] != 'Established'
+    ]
+
+    if protocols_down:
+        return (ExitCodes.critical, 'Some protocols are down!')
+    else:
+        return (ExitCodes.ok, 'All protocols are up')
+
+
+def get_birdc_protocols(path):
+    bird = Popen([path, 'show', 'protocols', ], stdout=PIPE, stderr=STDOUT)
```

Comment:
How likely is it that we will extend/change the list here?
If not likely, please go for:
```suggestion
    bird = Popen([path, 'show', 'protocols'], stdout=PIPE, stderr=STDOUT)
```

Reply:
That slid through from when I moved away from a multi-line. 

It's fixed in aa67e92.

Reply:
That slid through from when I moved away from a multi-line.

It's fixed in aa67e92.